### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by noraj** [notable/notable#663]
### Feature description

It seems notable has no documentation:

- **User documentation**: to explain the user how to use the app
- **Developer documentation**: to explain to devs the architecture of the code, how to dev (git practices, how to build release for all platforms, lint/style to respect, etc)
- **Code documentation**: for describing the code, mostly for dev

### Feature motivation

Imagine you are a dev and want to contribute: "Oh this project looks cool, I may contribute, where is the dev documentation? Source code is not commented / there is no code documentation? Forget it I won't contribute it will be too much to understand the main dev mind without doc."

Personally I came here and wanted to know how to package the tool for several Linux distributions, and how to build the binaries from source. No tar.gz for general linux purpose? How can I build the AppImage from source?

And a user can ask himself, how can I install this app for my distro ?

### Idea

For **User documentation** and **Developer documentation**: docusaurus, mkdocs, docsify, etc.

**Code documentation**: JSDoc, ESDoc, YUIDoc, etc.

### Start

There is the CONTRIBUTING guide for dev: https://github.com/notable/notable/blob/master/.github/CONTRIBUTING.md

There is the tutorial for users: https://github.com/notable/notable/tree/master/resources/tutorial/notes

**Comment by Armagedoom** [comment]
There is a tool that helped me tons but I discovered too late, which is the developer tools under the "help" menu.
Using those tools, you can set break points, see the variables, search the code for specific lines of code, and all you would require for developing.
This information should be added somewhere in the CONTRIBUTING.md in my opinion .

**Comment by fabiospampinato** [comment]
> User documentation

We have the tutorial notes for that. There should be a more easily accessible online version of them though.

> Developer documentation
> Code documentation

There's a lot to document here. Just documenting [Svelto](https://github.com/svelto/svelto), one of the projects used within the app, would probably take me multiple months. As of right now I think it is better for me to spend my time directly improving Notable instead.

Also in the future the core code of the app will probably be no longer released as open-source: https://github.com/notable/notable/issues/432, and instead most of the functionalities will be provided by plugins, when that happens we surely need some good documentation on how to develop plugins.

Touching some of the specific points you guys mentioned:

> git practices

I wouldn't know what to write about that really.

> how to build release for all platforms

This isn't documented, but if you've done any JS development you'll surely be familiar with NPM scripts, just by taking a look at those you'll see that there are multiple scripts called `build:*`.

>  lint/style to respect

This is a pretty important issue, but I've used jslint and the likes in the past and I mostly ended up fighting with the linter rather than getting useful info out of it. I think in the future I'll have to configure TSLint properly, probably by also writing some custom plugins for it. Or I should find some kind of formatter that I like. This is definitely a problem I want to tackle at some point.

> Source code is not commented

It's definitely not commented compared to much bigger/stable projects or compared to educational projects, but there are some comments here and there where I feel they are needed. As the project is still in its early states I think it's probably unwise to spend too much time around this, as the codebase will change significantly in the future.

> No tar.gz for general linux purpose? How can I build the AppImage from source?

Check the `build:linux` script. Even just searching for "AppImage" in the project you can see where we are telling `electron-builder` to build the `AppImage` version of the app. `electron-builder` itself is documented, in case you want to use some other build targets.

> how can I install this app for my distro ?

Until we make a proper website #481 this project will be a bit unapproachable for people not familiar with GitHub, but for developers I think it's relatively common practice to check the "release" page of the repo, and we have a big "DOWNLOAD" link right at the top of the readme. If you meant it more like "I'm using Ubuntu, how do I use this thing?" that will be made easier once we publish the app into the stap store #98 for instance and once we have a website, generally as of today you're kind of supposed to be able to install a `.deb` on your computer.

**Comment by noraj** [comment]
> We have the tutorial notes for that.

That's what I said.

> > git practices
> 
> I wouldn't know what to write about that really.

Like if dev must commit into a dev branch, how your release cycle is working that kind of things.





> This isn't documented, but if you've done any JS development you'll surely be familiar with NPM scripts, just by taking a look at those you'll see that there are multiple scripts called `build:*`.

I'm not a JS dev and that was not obvious to me. And yes this is part of the dev doc.

>  I think it's probably unwise to spend too much time around this, as the codebase will change significantly in the future.

So only you can easily understand the code, and it will be harder for external contributors but since you planned to close source the code anyway...

Also any project even little should have documentation, unit test, CI/CD, packaging, etc. here is a state of the art project I made in ruby https://gitlab.com/noraj/nvd_api with all of that.

You write comments like that in your code:

![image](https://user-images.githubusercontent.com/16578570/57880539-4b0d6880-781f-11e9-8923-4ae6e0f3b35a.png)

And it turns into code documentation like that after:

![image](https://user-images.githubusercontent.com/16578570/57880577-611b2900-781f-11e9-8304-556ebfbac3fd.png)

This is absolutely required for any project of that size.

> Check the `build:linux` script. Even just searching for "AppImage" in the project you can see where we are telling `electron-builder` to build the `AppImage` version of the app. `electron-builder` itself is documented, in case you want to use some other build targets.

If I would have known nodejs better I should have remember that all was in `package.json`. But it wasn't that obvious to me. Instead I looked in the readme or if I found documentation elsewhere (that didn't exist). That's normal reflex, not openning `package.json` except for devs.

> > No tar.gz for general linux purpose?

So? Usually a project provide and exe or msi installer for windows and a zip. And for Linux distro gives several packages for most popular distro or directly add and maintain them in official distro repos. And gives a generic tarball so other distro can build their own package. AppImage is just a big blob thatr won't take advantage of a linux FS organization and dependencies systems, it is just a big standalone binary that is including all dependencies, hopefully very few project do that else we would have the same dependencies hundred times.

![image](https://user-images.githubusercontent.com/16578570/57881079-a855e980-7820-11e9-80fc-3d8036051107.png)

> how can I install this app for my distro ?

I wasn't talking for me but you know not all users are techies, specially people using Ubuntu they are used to click on a *Install* button on Ubuntu store like on Google play store and that works not using dpkg or anything else with a manually downloaded package so giving them proper steps on how to do it will sure help them.

---

I'll be rude: To me it sounds like you're looking for excuses to avoid writing doc, any good IT pro known it is critical. Even if you want to work alone on the project it can help you since you can't remember all.

**Comment by fabiospampinato** [comment]
> This is absolutely required for any project of that size.

Evidently it isn't, or this project couldn't exist or something.

> If I would have known nodejs better I should have remember that all was in package.json. But it wasn't that obvious to me.

Even if we had top-notch documentation this is not the place where to learn about the JS ecosystem anyway.

> So?

So we are using `electron-builder` for building, if there's a way to make a tar.gz with it you'll find it in their docs, I don't know how to do that. 

> AppImage is just a big blob thatr won't take advantage of a linux FS organization and dependencies systems, it is just a big standalone binary that is including all dependencies, hopefully very few project do that else we would have the same dependencies hundred times.

This project is built around Electron, basically we are shipping a whole browser alongside the app, there isn't a way to compile to app into a bundle that dynamically links to the existing browser installation or something.

> I'll be rude: To me it sounds like you're looking for excuses to avoid writing doc, any good IT pro known it is critical. 

Yeah that was rude. I guess then I'm not an IT pro, or I'm not good. Please keep these sorts of comments to a minimum, they are personal, don't help improve the app, and remember that I don't owe you or anyone else anything, if you don't like how things are being run I'd be happy to hear your feedback about that, but ultimately I decide how to run this repo, if that's not ok with you you're absolutely free to fork it, you're welcome.

**Comment by noraj** [comment]
> > This is absolutely required for any project of that size.
> 
> Evidently it isn't, or this project couldn't exist or something.

What kind of bad faith is that? You know I meant it is very much recommended but big project without code doc, dev doc, unit test and devops process are vowed to fails or being un-maintainable.

> Even if we had top-notch documentation this is not the place where to learn about the JS ecosystem anyway.

I'm not aking you to write the whole JS doc, just tell somewhere in a dev doc "build scripts are in package.json, to know more how it works see this doc <with a link to external doc here>" or that kind of stuff instead of nothing.

> This project is built around Electron, basically we are shipping a whole browser alongside the app, there isn't a way to compile to app into a bundle that dynamically links to the existing browser installation or something.

So why Rambox, Atom, VScode, and all Electron projects are able to provide a tar.gz except yours ?

> Yeah that was rude. I guess then I'm not an IT pro, or I'm not good.

You sincerely think doc is optional in IT projects?... 

I'm quite disappointed, I first thought Notable would be the right and great answer to QOwnNotes, Vnotes, Boostnotes, Joplin but seems it is just a project maintained by a person alone that is close minded, stubborn and that doesn't want to learn from others.

With all the bad faith in the world you're telling me that doc is not important, that code doc is useless, that unit test is not avoiding bug and regressions, etc. Ok keep it that way, I'm actually helping on a project ruled by someone like you:

![image](https://user-images.githubusercontent.com/16578570/57883235-eacdf500-7825-11e9-8c55-b6c5acbd57e7.png)

Checkup: 2 years after I joined nothing has changed, the project is not maintainable, it is hard to deploy, people open dozen of bug issues every weeks and only less than 5% are fixed, the project is a giant mess. The only difference is this guy knowns it is important but is just to lazy and worried abotu others stuff than his own project where you are just relentlessly saying this is useless.

**Comment by fabiospampinato** [comment]
@noraj Don't put words in my mouth. I haven't said that docs or unit tests are useless. You've expressed your point, let's leave it at that. In the future please avoid writing these sorts of personal comments.

**Comment by noraj** [comment]
TL;DR; I opened an issue to kindly say with motivation and arguments why a documentation is required and you just took all point one by one and said, this isn't necessary I won't do it, this isn't necessary I won't do it, this isn't necessary I won't do it, this isn't necessary I won't do it.

So ok do nothing, let's close this and forget all of that.

**Comment by noraj** [comment]
Just for humor "seems to be chronic":

![image](https://user-images.githubusercontent.com/16578570/57884749-b1978400-7829-11e9-86dd-d3a81e1080e2.png)

**Comment by fabiospampinato** [comment]
Sorry @noraj, these kinds of comments are welcomed here. 

I'm reopening this as the documentation, especially user-oriented in the near term, is a point we should improve on.

**Comment by fabiospampinato** [comment]
As Notable, for now at least (?), is closed-source, I'm closing this public-facing issue.
